Functions/Mock.ps1

```diff
@@ -1357,6 +1261,155 @@ function Get-DynamicParamBlock {
+
+                    $code = @"
+                        $cmdletBinding
+                        param(
+                            [object] `${P S Cmdlet}$comma
+                            $paramBlockSafeForDynamicParams
+                        )
+
+                        `$PSCmdlet = `${P S Cmdlet}
+
+                        $dynamicParamStatements
+"@
+
+                    $dynamicParamScriptBlock = [scriptblock]::Create($code)
+
+                    $sessionStateInternal = Get-ScriptBlockScope -ScriptBlock $contextInfo.Command.ScriptBlock
+
+                    if ($null -ne $sessionStateInternal) {
+                        Set-ScriptBlockScope -ScriptBlock $dynamicParamScriptBlock -SessionStateInternal $sessionStateInternal
+                    }
+                }
+            }
+        }
+
+        $EscapeSingleQuotedStringContent = if ($global:PSVersionTable.PSVersion.Major -ge 5) {
+            { [System.Management.Automation.Language.CodeGeneration]::EscapeSingleQuotedStringContent($args[0]) }
+        }
+        else {
+            { $args[0] -replace "['‘’‚‛]", '$&$&' }
+        }
+
+        $newContent = & $SafeCommands['Get-Content'] function:\MockPrototype
+        $newContent = $newContent -replace '#FUNCTIONNAME#', (& $EscapeSingleQuotedStringContent $CommandName)
+        $newContent = $newContent -replace '#MODULENAME#', (& $EscapeSingleQuotedStringContent $ModuleName)
+
+        $canCaptureArgs = 'true'
+        if ($contextInfo.Command.CommandType -eq 'Cmdlet' -or
+            ($contextInfo.Command.CommandType -eq 'Function' -and $contextInfo.Command.CmdletBinding)) {
+            $canCaptureArgs = 'false'
+        }
+        $newContent = $newContent -replace '#CANCAPTUREARGS#', $canCaptureArgs
+
+
+        $functionBody = @"
+        $cmdletBinding
+        param(
+            $paramBlock
+        )
+        $dynamicParamBlock
+        begin
+        {
+            `${mock call state} = @{}
+            $($newContent -replace '#BLOCK#', 'Begin' -replace '#INPUT#')
+        }
+        process
+        {
+            $($newContent -replace '#BLOCK#', 'Process' -replace '#INPUT#', '-InputObject @($input)')
+        }
+        end
+        {
+            $($newContent -replace '#BLOCK#', 'End' -replace '#INPUT#')
+        }
+"@
+
+        $functionName = [Guid]::NewGuid().Guid
+        $code = @"
+            begin {
+                `$script:receivedArgs = `$args
+                function Internal_$functionName {
+                    $functionBody
+                }
+            }
+            end {
+                `$receivedInput = @(`$input)
+
+                if (`$receivedInput.Count -gt 0) {
+                    `$receivedInput | Internal_$functionName @args
+                }
+                else {
+                    Internal_$functionName @args
```

Comment:
@nohwnd I think I found part of the issue with PowerShell 2. It seems to be that splatting the `$args` variable to the new internal mock function does not work properly in PS 2.

I inserted a ```Write-Host `$(`$PSBoundParameters | Out-String)``` in the dynamicParam block to print the static parameters resolution using:

```
Mock Set-Item { return "I am not Set-Item" }
Set-Item "mypath" -value "value"
```

The invocation prints to console:
- PS 5.1 : 
Key   Value
---   -----
Value value
Path  {mypath}

- PS 2:
Key                                                                 Value
---                                                                 -----
Path                                                                {mypath}
Value                                                               -value

So, in PS 2 it is trying to map to the *-value* string as the input for the 
**-Value**. I suppose this is due to **-Value** being a positional parameter

Reply:
Based on this, we may want to consider merging this for pester v5 instead of master.

Thoughts?